### DISCUSSION TITLE
IntelliJ Plugin for generating some of the boiler plate in a store?

### DISCUSSION THREAD

**Discussion by plusmobileapps** [arkivanov/MVIKotlin#34] [Ideas]
I saw in one of the latest releases the [experimental DSL](https://arkivanov.github.io/MVIKotlin/store.html) for creating a store which was a great improvement. Although it can still be a little tedious to create a store and was thinking to help speed up the creation it would be nice to have an intellij plugin that could auto generate the basic structure of a store similar to the [flutter bloc](https://plugins.jetbrains.com/plugin/12129-bloc) plugin? 

I was thinking it could just be an option in the context menu to "create new MVIKotlin store" which could open a dialog with the following options: 

* radio group for which executor you are using (coroutines or reaktive)
* checkbox if you intend on using the `Label`
* text box for the name of the store
* text box for the type of bootstrapper if one is needed

Then based off all that could just generate a couple files with the skeleton for the developer to fill out.

```kotlin
interface FooStore : Store<Intent, State, Nothing> {

    sealed class Intent {
            // TODO create intents
    }

    data class State(
          // TODO create initial state
    )

}

class FooStoreProvider(private val storeFactory: StoreFactory) {

    sealed class Message {
             // TODO create messages that update store state
    }

    fun provide(): FooStore = object : FooStore,
        Store<Intent, State, Nothing> by storeFactory.create<Intent, Unit, Message, State, Nothing>(
            name = "FooStore",
            initialState = State(),
            bootstrapper = bootstrapper(),
            executorFactory = createExecutor(),
            reducer = onMsgReceived()
        ) {}

    private fun onMsgReceived(): State.(msg: Message) -> State = { msg ->
        TODO()
    }

    private fun createExecutor(): () -> Executor<Intent, Unit, State, Message, Nothing> =
        reaktiveExecutorFactory {
            onIntent<Intent> { intent ->
                // TODO
            }
        }

    private fun bootstrapper(): Bootstrapper<Unit> = reaktiveBootstrapper {
        // TODO 
    }
}
```

Curious what you think of this idea?

**Comment by arkivanov** [comment]
Yes, this could be added. Wondering what would be a better way - to add this feature to the existing time travel plugin, or to introduce a separate one. Both ways have their pros and cons. I would to hear some feedback from the community.

**Reply by plusmobileapps** [reply]
Hmm...if it were to be added to the time travel plugin it might make sense to rename it mvi kotlin plugin so it could encompass all plugin features. Otherwise could split them up. 

I might lean towards having one unified plugin since these are all tools I would use anyways if I were to us the library.

**Reply by arkivanov** [reply]
Yeah, seems like updating the existing plugin is a good idea! There could be also a dialog with options:

- Store name
- Implementation variant - classic or DSL
- Factory variant - a class or a top-level function
- Checkboxes for Intent, State and Label

**Comment by malliaridis** [comment]
I was about to create a new discussion about file templates but found this one and thought it fits well in here. I had similar ideas about choosing the executor and stuff like that, but I kept things simple because I have no clue how to implement that (except with different templates).

So I created a file template for IntelliJ IDEA that can be used for quickly create the classes.

The current set does not contain `ChildStack` / routing functionality from Decompose, but I plan to add another template for such components. The goal is to have at the end commented and uncommented file templates for components with and without `ChildStack` (this one is without `ChildStack` and with comments), plus the option between Kotlin coroutines (this one is based on coroutines) and Reaktive.

https://user-images.githubusercontent.com/33470669/189489308-eebf4ac1-61a2-44e3-a7fc-47ae789163a3.mov

The templates can be imported into a project by simply placing the files from [idea.zip](https://github.com/arkivanov/Decompose/files/9540492/idea.zip) into the projects `.idea/` directory. Not the best way of integrating stuff like that, but better than nothing I guess. A plugin would very likely be a great extension though.

EDIT: There is only the `.asValue()` function that is not covered by the templates, because it is a custom function (not provided yet by the library).
EDIT 2: After the discussion about [Naming Conventions](https://github.com/arkivanov/Decompose/discussions/201#discussioncomment-3614940) I have to update the class names.

**Reply by arkivanov** [reply]
This looks useful, thanks for sharing!

**Comment by arkivanov** [comment]
Oh no, I totally forgot about this discussion! I have filed #61 to investigate it further (and probably create something).

**Reply by arkivanov** [reply]
There are live templates available to download. See "IDEA Live Template" section in the [docs](https://arkivanov.github.io/MVIKotlin/store.html). Wondering, do we still need this in the IDEA plugin?

**Reply by malliaridis** [reply]
I think live templates are more sustainable than a plugin. I experience on every IDE update warnings that some plugins are not supporting the current version, so I think it would be a burden to keep the plugin up-to-date. And EAP versions are often not supported at all because of the frequent updates.

So I think live templates is the way to go here.
